src/line_launcher/mod.rs

```diff
@@ -403,13 +450,23 @@ impl LineLauncher {
                             next_note_index,
                             SystemTime::now()
                         );
+                        let random_ratio_of_duration_between_sixteenth_notes =
+                            if let Some(randomize_note_start_time_ratio) =
+                                randomize_note_start_time_ratio
+                            {
+                                (thread_rng.gen::<f64>() - 0.5) * randomize_note_start_time_ratio
+                            } else {
+                                0.0
+                            };
                         schedule_note_on_sender
                             .send(ScheduleNoteOnMessage {
                                 time: SystemTime::now()
                                     + duration_between_sixteenth_notes.mul_f64(
-                                        1.0 + ((0.5
+                                        (1.0 + ((0.5
                                             - ahead_or_behind_the_beat_ratio.unwrap_or(0.5))
-                                            * 2.0),
+                                            * 2.0)
+                                            + random_ratio_of_duration_between_sixteenth_notes)
+                                            .max(0.0),
```

Comment:
This could take us into an overall negative ratio here between being ahead of the beat + randomness so cap it at zero to avoid a panic